### PR TITLE
[mlir][Linalg] Add speculation for LinalgStructuredOps

### PR DESCRIPTION
**Pull request by Groverkss**
This patch adds speculation behavior for linalg structured ops, allowing them to be hoisted out of loops using LICM.

### CODE REVIEW

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1212]:
```suggestion
  // The body of the op can still have speculation in its region.
```

File: mlir/test/Transforms/loop-invariant-code-motion.mlir
Comment (kuhar) [inline mlir/test/Transforms/loop-invariant-code-motion.mlir:1165]: Could we have a test with something to exercise the recursive aspect of speculablity? Maybe potential division by zero in the body?

Comment (Groverkss) [reply mlir/test/Transforms/loop-invariant-code-motion.mlir:1165]: added

**Review by kuhar — APPROVED** [review]
LGTM but let's wait for an approval from someone from the linalg side

**Review by dcaballe — COMMENTED** [review]
This is cool :)

I was wondering if, in addition to the tensor vs memref semantics, we should take into account the speculatability of the nested operations?

**Comment by Groverkss** [comment]
> This is cool :)
> 
> I was wondering if, in addition to the tensor vs memref semantics, we should take into account the speculatability of the nested operations?

I think we are accounting that. On tensor semantics, the speculatability of linalg ops is defined as recursively speculatable. I checked licm implementation, and it recursively checks every region if an op is recursively speculatable, so this should work. There is an example test added which does not hoist because a nested op could cause division by zero.

**Review by dcaballe — APPROVED** [review]
Awesome!

>  division by zero.

That's exactly the case I was thinking about. Sorry, I missed the CHECKs.

**Review by MaheshRavishankar — APPROVED** [review]
I don't know much about this but looks mostly right to me.

**Review by rengolin — APPROVED** [review]

**Pull request merged**
